### PR TITLE
[Fix #3252] Remove dependency on react-native-autolink

### LINKED ISSUE #3252: Remove dependency on react-native-autolink
### User Story

As a developer, I want to rely on dependency only when mandatory

### Description

*Type*: Feature

`react-native-autolink` is used in a [single place](https://github.com/status-im/status-react/blob/80fb8dde8baea957d37cdf94f0ea88f74b7c810a/src/status_im/chat/views/message/message.cljs#L142) to detect URL and give them proper styling.
This is trivial to do with a regexp, not need to rely on an external dependency for this.

Each dependency is an extra security concern.

### PR DESCRIPTION
**Pull request by foopang**
fixes #3252 

### Summary:

 Remove dependency on react-native-autolink

status: ready

### CODE REVIEW

File: src/status_im/chat/views/message/message.cljs [hunk 1]
```diff
@@ -111,13 +114,39 @@
...
+
 ;; todo rewrite this, naive implementation
-(defn- parse-text [string]
+(defn- parse-text [string url-on-press]
   (if (string? string)
     (let [general-text  (string/split string regx)
           general-text' (if (zero? (count general-text))
                           [nil]
                           general-text)
+          general-text' (vec (map-indexed (fn [idx string]
```
Comment (jeluard) [inline src/status_im/chat/views/message/message.cljs]: Can we came up with a more descriptive var name?
Also let's extract this logic as a private function.

File: src/status_im/chat/views/message/message.cljs [hunk 2]
```diff
@@ -111,13 +114,39 @@
        ffirst
        replacements))
 
+(defn- parse-url [string on-press]
```
Comment (jeluard) [inline src/status_im/chat/views/message/message.cljs]: This looks like a good candidate for unit testing!
Also could you simplify it a bit by introducing some private helper functions?

**Review by jeluard — CHANGES_REQUESTED** [review]
Great job, thanks!
I added some comment around code style.

File: src/status_im/chat/views/message/message.cljs (hunk 2 above)
Comment (janherich) [reply src/status_im/chat/views/message/message.cljs]: I agree that unit tests would be super nice for that function !

File: src/status_im/chat/views/message/message.cljs (hunk 1 above)
Comment (janherich) [reply src/status_im/chat/views/message/message.cljs]: I'm also not big fan of `symbol'` etc. I would rather have one binding with combined expression (using `->`/`->>` threading macros), or more descriptive names (such as `general-text-seq`,  `general-text-react-seq`, describing each step of the transformation process)

**Review by janherich — CHANGES_REQUESTED** [review]

**Comment by foopang** [comment]
@jeluard @janherich Updated. Looks better now?

File: src/status_im/chat/views/message/message.cljs [hunk 3]
```diff
@@ -112,37 +144,37 @@
...
-                                                :style style}
-                                               (subs string 1 (dec (count string)))]))
-                                          (re-seq regx string)))
-          styled-text'  (if (> (count general-text)
-                               (count styled-text))
-                          (conj styled-text nil)
-                          styled-text)]
-      (mapcat vector general-text' styled-text'))
+    (let [general-text (string/split string regx)
+          general-text (if (zero? (count general-text))
```
Comment (jeluard) [inline src/status_im/chat/views/message/message.cljs]: Can you extract those code chunks as standalone function with proper name? Then thread them using `->`.
This way the code is clearer and no need to reuse var names.

Same for the other large function.

Comment (foopang) [reply src/status_im/chat/views/message/message.cljs]: @jeluard Updated.

File: src/status_im/chat/views/message/message.cljs [hunk 4]
```diff
@@ -112,37 +147,41 @@
...
+    (letfn [(create-general-text [text-seq]
+              (if (zero? (count text-seq))
+                [nil]
+                (->> text-seq
+                     (map-indexed (fn [idx string]
+                                    (apply react/text
+                                           {:key (str idx "_" string)}
+                                           (autolink string url-on-press))))
+                     vec)))]
+      (let [general-text    (-> string
```
Comment (jeluard) [inline src/status_im/chat/views/message/message.cljs]: This chunk is duplicated in `parse-url`

File: src/status_im/chat/views/message/message.cljs [hunk 5]
```diff
@@ -112,37 +147,41 @@
...
+            styled-text-seq (->> string
+                                 (re-seq regx)
+                                 (map-indexed (fn [idx string]
+                                                (let [style (get-style string)]
+                                                  [react/text
+                                                   {:key   (str idx "_" string)
+                                                    :style style}
+                                                   (subs string 1 (dec (count string)))])))
+                                 vec)
+            styled-text     (cond-> styled-text-seq
```
Comment (jeluard) [inline src/status_im/chat/views/message/message.cljs]: This chunk is duplicated in `parse-url`

File: src/status_im/chat/views/message/message.cljs [hunk 6]
```diff
@@ -102,6 +103,40 @@
 
 (def regx (re-pattern (string/join "|" (map first replacements))))
 
+(def regx-url #"https?://[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]")
+
+(defn parse-url [string]
+  (letfn [(create-non-url-text [text-seq]
```
Comment (jeluard) [inline src/status_im/chat/views/message/message.cljs]: Why a `letfn` and not a top level `defn` ?

**Review by jeluard — CHANGES_REQUESTED** [review]
@foopang Thanks for the changes!

Actually what I meant is introduce shorter top level `defn` that will be reused in `parse-text` and `parse-url`.

This way the code is easier to read and you can remove code duplication.

**Comment by foopang** [comment]
@jeluard I see what you mean now. Updated and added `parse-str-regx` reusable for both `parse-text` and `parse-url`. Please take a look.

**Comment by yenda** [comment]
@jeluard and @janherich have your concerned been addressed ?
@foopang can you resolve conflicts and squash your commits ?

File: src/status_im/chat/views/message/message.cljs [hunk 7]
```diff
@@ -100,49 +101,83 @@
...
    "~[^~]+~"     {:font-style :italic}})
 
-(def regx (re-pattern (string/join "|" (map first replacements))))
+(def regx-styled (re-pattern (string/join "|" (map first replacements))))
+
+(def regx-url #"https?://[-a-zA-Z0-9+&@#/%?=~_|!:,.;]*[-a-zA-Z0-9+&@#/%=~_|]")
+
+(defn- parse-str-regx [string regx matched-fn unmatched-fn]
+  (if (string? string)
+    (letfn [(create-unmatched-text [text-seq]
```
Comment (jeluard) [inline src/status_im/chat/views/message/message.cljs]: Any reason to have a `letfn` and not a `defn` ?

Comment (foopang) [reply src/status_im/chat/views/message/message.cljs]: @jeluard Removed `letfn` now as what it does is just a simple condition check, can be accomplished using `as->` and `if` combined. What do you think?

**Comment by jeluard** [comment]
@foopang Awesome thanks again!

**Review by jeluard — APPROVED** [review]

**Review by janherich — APPROVED** [review]

**Comment by status-open-bounty** [comment]
Contract deploying
Transaction: [0x75a7c5e6514cd60ab5f7d914774f2934a04c1588b8e41a6cab92b4e02a678ac5](https://etherscan.io/tx/0x75a7c5e6514cd60ab5f7d914774f2934a04c1588b8e41a6cab92b4e02a678ac5)

**Comment by lukaszfryc** [comment]
@foopang here are some issues I've found while testing your changes on iOS and Android: 

1. iOS - URLs that do not start with "http" are not shown as links

 

Please make sure that following URLs are shown as "google.com" link:
- `www.google.com` (currently do not work)
- `google.com` (currently do not work)
- `http://www.google.com`
- `http://google.com`

2. iOS - Long URLs are not collapsed



Android works fine. Here is how it looks like:



3. Android - There are some UI glitches (white rectangles) in DApp profile screen 



To reproduce: 
1. Go to Home
2. Click "+" and Open DApp
3. Open "Auction House" DApp

**Review by lukaszfryc — DISMISSED** [review]
@foopang please check the issues in my latest comment.

**Comment by foopang** [comment]
@lukaszfryc The above issues have been fixed now. Please kindly check this again. Thanks.

**Comment by lukaszfryc** [comment]
@jeluard could you have a look at the latest changes? In the meantime, I'll test the app.

**Comment by jeluard** [comment]
@foopang Do I understand right that you updated the logic to detect web links? If so would you mind adding one test or two to reflect this?

**Comment by lukaszfryc** [comment]
@foopang the last issues are fixed. But, I've also noticed that we create links out of values like "$33.90". It does not make sense and since such messages may appear in chats quite often we should fix it. @foopang probably a simple rule like checking if a text matches IP address or not is enough. 
 
`react-native-autolink` has other rules like skipping links for incorrect domains ("google.com" is a link but "google.ss" is not) or creating mailto links for email addresses. But, I think we should not implement them in this PR. Let's only fix the first issue with decimal point ("33.9").

**Comment by foopang** [comment]
@lukaszfryc @jeluard I have fixed it now, and updated test to reflect the changes.

**Comment by jeluard** [comment]
@foopang Awesome thanks!

@lukaszfryc Please remove your review (this is supposed to be for developers)

**Comment by lukaszfryc** [comment]
@foopang looks good on my side as well. Thank you!

**Pull request merged**